### ISSUE TITLE
feat(layoutlm): Python 3.14 + experimental Core AI (.aimodel) export for Swift

### ISSUE THREAD

**Issue by tnorlund** [tnorlund/Portfolio#1680]
## Goal

Modernize the receipt LayoutLM stack so the normal Python training/inference environment can move to **Python 3.14 + current PyTorch**, while preserving the existing Core ML exporter and adding an **experimental Apple Core AI `.aimodel` export/runtime path** for Swift on macOS 27 / Xcode 27.

Do **not** replace the working Core ML path yet. Core AI should land alongside it behind an explicit opt-in until conversion + numerical + Swift runtime parity are proven.

## Why now

As of 2026-09-17:

- PyTorch supports Python 3.14 on macOS.
- Hugging Face Transformers/LayoutLM can run on current Python/PyTorch; LayoutLM itself is no longer the primary blocker.
- Apple released Core AI / `coreai-torch`, which converts `torch.export.ExportedProgram` into Core AI assets for Swift/on-device execution.
- The current Apple Core AI converter stack still has a Python-version mismatch: `coreai-torch` supports modern PyTorch, but the binary `coreai-core` dependency currently does not provide Python 3.14 support, so Core AI export should remain in an isolated Python 3.13 environment for now.
- Existing `coremltools 9.0` compatibility is also tied to the older PyTorch range, so the legacy Core ML exporter should stay isolated instead of constraining the whole package.

References:
- https://pytorch.org/get-started/locally/
- https://pytorch.org/blog/pytorch-2-10-release-blog/
- https://huggingface.co/docs/transformers/installation
- https://pypi.org/project/coreai-torch/
- https://github.com/apple/coreai-torch
- https://pypi.org/project/coreai-core/
- https://github.com/apple/coreai-models

## Current repo state

The repo already has a mature Core ML + Swift path.

Relevant files:

- `receipt_layoutlm/pyproject.toml`
  - `requires-python = ">=3.13"`
  - currently pins shared `torch>=2.6.0,<=2.7.0` specifically so `coremltools 9.0` can export checkpoints.
  - this shared pin is now the main thing preventing the ordinary project environment from taking advantage of Python 3.14/current PyTorch.
- `receipt_layoutlm/receipt_layoutlm/export_coreml.py`
  - wraps LayoutLM/LayoutLMv3
  - uses TorchScript + `coremltools`
  - exports `LayoutLM.mlpackage`
  - packages tokenizer/config/label map.
- `receipt_layoutlm/receipt_layoutlm/validate_coreml.py`
  - compares PyTorch vs Core ML labels, confidences and logit RMSE.
- `receipt_layoutlm/receipt_layoutlm/export_worker.py`
  - macOS SQS worker
  - downloads the checkpoint from S3
  - exports
  - publishes versioned + canonical bundles
  - writes model provenance.
- `.agents/skills/coreml-export/SKILL.md`
  - documents the isolated `~/.coreml-venv` flow.
- `receipt_ocr_swift/Sources/ReceiptOCRCore/LayoutLM/LayoutLMInference.swift`
  - currently imports CoreML
  - compiles/loads `.mlpackage`
  - owns tokenizer, bbox construction, windowed inference, v1/v3 image handling, softmax, and per-word aggregation.
- `.agents/skills/mac-ocr-worker/SKILL.md`
  - documents the Swift Vision OCR + LayoutLM worker and model cache.
- Existing model bundle metadata should remain reusable:
  - tokenizer files / `vocab.txt`
  - `config.json`
  - `label_map.json`
  - `model_identity.json`.

Important behavioral invariants:

- Current production behavior is LayoutLM token classification over Apple Vision OCR words + normalized bboxes.
- Swift windowed inference must continue matching training defaults (`windowSize=200`, `windowStride=150`).
- First-subtoken supervision/aggregation must not change.
- v1 uses WordPiece; v3 uses BPE + optional `pixel_values`.
- Agents must use the **dev** stack only; do not touch prod.

## Proposed architecture

Use three deliberately separate Python environments.

### 1. Main LayoutLM environment — Python 3.14

Used for normal development/training/inference.

Target:

```
Python 3.14
current supported PyTorch
Transformers
receipt_layoutlm[training,dev]
```

This environment should **not** be constrained by `coremltools`.

### 2. Legacy Core ML exporter — Python 3.13

Preserve the current exporter:

```
Python 3.13
torch 2.6–2.7
coremltools 9.0
receipt_layoutlm[coreml]
```

No functional regression.

### 3. Core AI exporter — Python 3.13 for now

New isolated exporter:

```
Python 3.13
torch >= 2.8
coreai-torch
coreai-core transitively
receipt_layoutlm[coreai]
```

This environment exists only because Apple's Core AI binary dependency has not caught up to Python 3.14 yet. Once Apple adds 3.14 support we should be able to collapse this into the main environment.

### A. Decouple PyTorch from the Core ML compatibility pin

Update `receipt_layoutlm/pyproject.toml` so the base package no longer globally caps PyTorch at 2.7.

A reasonable dependency shape is:

```toml
dependencies = [
  "torch>=2.6,<3",
  ...
]

[project.optional-dependencies]
coreml = [
  "torch<=2.7.0",
  "coremltools==9.0",
]

coreai = [
  "torch>=2.8.0",
  "coreai-torch",
]
```

The exact pins may be adjusted if dependency resolution or repo tests show a better combination, but preserve these invariants:

- normal Python 3.14 install resolves to a modern PyTorch
- `[coreml]` on Python 3.13 resolves to PyTorch <=2.7 + coremltools 9.0
- `[coreai]` on Python 3.13 resolves to PyTorch >=2.8 + Apple Core AI packages.

Add Python 3.14 project/classifier/test coverage where appropriate. Do **not** make 3.14t/free-threaded a requirement.

### B. Add a Core AI export module

Add something like:

`receipt_layoutlm/receipt_layoutlm/export_coreai.py`

Start with **LayoutLM v1**, which is the safest current production target. Preserve v3 Core ML behavior; add Core AI v3 only if it is straightforward and covered.

Reuse/refactor the existing wrapper so Core ML and Core AI do not silently diverge in their model inputs.

Expected v1 interface:

- `input_ids`
- `attention_mask`
- `bbox`
- `token_type_ids`
- output: `logits`

Prefer a **fixed sequence length of 512 for the first Core AI implementation** if dynamic sequence shapes materially complicate `torch.export`/Core AI conversion. Swift already pads/truncates to the model maximum, so fixed 512 is acceptable for the first parity implementation.

Conversion should follow Apple's currently supported flow, approximately:

```python
wrapper.eval()

ep = torch.export.export(wrapper, args=sample_inputs)
ep = ep.run_decompositions(coreai_torch.get_decomp_table())

converter = coreai_torch.TorchConverter()
converter.add_exported_program(
    ep,
    input_names=[...],
    output_names=["logits"],
)
program = converter.to_coreai()
program.optimize()
program.save_asset(...)
```

Use the actual installed Apple API rather than copying this pseudocode blindly.

Export a bundle such as:

```
model-bundle/
├── LayoutLM.aimodel/
├── vocab.txt
├── config.json
├── label_map.json
├── tokenizer_config.json   # when present
└── model_identity.json
```

Keep metadata/labels/tokenizer semantics compatible with the existing Swift worker.

### C. Add Core AI parity validation

Add a Core AI counterpart to `validate_coreml.py`, or refactor shared comparison logic so both backends use the same metric code.

Validation must compare:

- raw logits / RMSE
- argmax labels
- confidence deltas
- NaN/Inf detection
- label mapping consistency

Use deterministic inputs for a unit/smoke test, and support real receipt samples in the same spirit as the current Core ML validator.

Do not claim parity from converter success alone.

Where the runtime permits, validate:

1. PyTorch reference
2. Core AI CPU specialization/runtime
3. default Core AI specialization

If the cloud agent cannot execute Core AI runtime or Swift because it is not on a suitable Mac, it must still implement the code/tests it can run and document exact local verification commands. Do not fabricate successful runtime results.

### D. Add CLI support

Add explicit commands rather than overloading the existing command ambiguously. For example:

```bash
layoutlm-cli export-coreai ...
layoutlm-cli validate-coreai ...
```

Keep `export-coreml` unchanged.

Document creation of an isolated Core AI environment, e.g.:

```bash
/usr/local/bin/python3.13 -m venv ~/.coreai-venv
~/.coreai-venv/bin/pip install -e 'receipt_layoutlm[coreai]'
```

Also document a Python 3.14 main/dev environment.

### E. Add an opt-in Core AI Swift backend

Do not rewrite all receipt post-processing.

Refactor only enough to allow the model execution backend to vary while preserving:

- tokenization
- bbox alignment
- reading-order/window logic
- first-subtoken mapping
- softmax/label mapping
- OCR worker API/results

Suggested shape:

```
LayoutLMInference
    ├── shared preprocessing/postprocessing
    ├── CoreML backend     (existing/default)
    └── CoreAI backend     (new/opt-in)
```

Add a Core AI runner using the current Core AI Swift runtime API and gate it appropriately for macOS 27 / Xcode 27.

An explicit selection such as `LAYOUTLM_BACKEND=coreml|coreai` or an equivalent CLI/config flag is preferable for the initial landing.

**Default must remain Core ML** until parity has been demonstrated.

Avoid globally raising deployment requirements unless necessary. Prefer availability checks / conditional import so older Core ML builds can still compile when practical.

### F. Do not refactor the SQS/S3 export pipeline yet unless required

The first PR should prove:

- dependency separation
- Python 3.14 compatibility
- `.aimodel` export
- validation tooling
- Swift local loading/inference behind an opt-in backend

Do not rename `CoreMLExportJob`, change production S3 pointers, or replace the existing auto-export pipeline as part of this first change unless there is a strong technical reason.

Once Core AI parity is proven, follow-up work can add a parallel `coreai/` model pointer and automated export/promotion.

## Tests / acceptance criteria

- [ ] A fresh Python 3.14 environment can install the normal `receipt_layoutlm` package and its training/dev dependencies without the old `torch<=2.7` cap.
- [ ] Existing Python 3.13 `receipt_layoutlm[coreml]` resolves to the existing Core ML-compatible PyTorch/coremltools versions and current Core ML tests continue to pass.
- [ ] A Python 3.13 `receipt_layoutlm[coreai]` environment installs successfully with `coreai-torch` and a PyTorch version >=2.8.
- [ ] New `export-coreai` code can export a known LayoutLM v1 checkpoint to a `.aimodel` asset using `torch.export`.
- [ ] The Core AI bundle carries the same tokenizer/config/label/provenance data needed by the Swift worker.
- [ ] Core AI validation compares the same fundamental outputs as Core ML validation and fails clearly on NaN/Inf or major label disagreement.
- [ ] Swift has an opt-in Core AI backend while Core ML remains the default.
- [ ] Shared Swift windowing/tokenization/first-subtoken semantics are unchanged.
- [ ] Existing Swift/Core ML tests pass.
- [ ] Add targeted tests for the new backend/selection/config logic.
- [ ] Add/update agent documentation in `.agents/skills/coreml-export/SKILL.md` and `.agents/skills/mac-ocr-worker/SKILL.md` (or add a dedicated `coreai-export` skill if cleaner).
- [ ] If macOS 27/Core AI execution cannot be run in CI/cloud, the PR documents exact local commands and clearly marks which runtime checks still require a compatible Mac.
- [ ] No prod stack changes.

## Known risks / things to watch

- Core AI / `coreai-torch` is still young; converter/runtime issues should be treated as expected integration risk.
- The Apple Core AI binary dependency is currently the reason the export environment cannot simply be Python 3.14.
- Do not assume TorchScript behavior maps 1:1 to `torch.export`.
- LayoutLM relies heavily on integer token IDs/bboxes + embedding/gather behavior; verify integer semantics numerically.
- Start with float32 for parity if necessary before optimizing precision.
- Do not introduce quantization until the float path is proven.
- Preserve the existing Core ML path as the fallback.

## Definition of done

A PR is ready when the repo can use **Python 3.14 for normal LayoutLM work without breaking the isolated Core ML exporter**, and it contains a real, opt-in **Core AI LayoutLM v1 path from PyTorch checkpoint → `.aimodel` → Swift backend**, with parity tooling and honest documentation of any macOS-only validation that still needs to be run locally.
